File: Python MATHS/finite_field_arithmetic.py
####################
# Máximo común divisor:
# ax + by = 1
####################

def xgcd(x, y):
    old_r, r = (x, y)
    old_s, s = (1, 0)
    old_t, t = (0, 1)
    count = 0
    while r != 0:
        print("\tEncontrando el xgcd({}): {} {} {}".format(count, old_r, old_t, old_s), end="\r")
        count += 1
        quotient = old_r // r
        old_r, r = (r, old_r - quotient * r)
        old_s, s = (s, old_s - quotient * s)
        old_t, t = (t, old_t - quotient * t)

    print()
    return old_s, old_t, old_r  # a, b, g

primo = 13

print("Operadores matemáticos en campo finito:")
####################
# Adición en campo finito
####################
izquierda = 4
derecha = 12
print("\tSuma: ({} + {}) % {} = {}\n".format(izquierda, derecha, primo, (izquierda + derecha) % primo))

####################
# Sustracción en campo finito
####################
izquierda = 35
derecha = 5
print("\tResta: ({} - {}) % {} = {}\n".format(izquierda, derecha, primo, (izquierda + derecha * -1) % primo))

####################
# Multiplicación en campo finito
####################
izquierda = 90
derecha = 10
print("\tMultiplicación: ({} * {}) % {} = {}\n".format(izquierda, derecha, primo, (izquierda * derecha) % primo))

####################
# División en campo finito
####################
dividendo = 2
divisor = 3
print("\tDivisión: ({} / {})".format(dividendo, divisor))
a, b, g = xgcd(divisor, primo)
print("\tse encontró: {}*{} + {}*{}=1".format(a, divisor, b, primo))
print("\t({} / {}) % {} = {}\n".format(dividendo, divisor, primo, (dividendo * a) % primo))

File: Rust MATHS/src/finite_field_arithmetic.rs
fn xgcd(x: i32, y: i32) -> (i32, i32, i32) {
    let (mut old_r, mut r) = (x, y);
    let (mut old_s, mut s) = (1, 0);
    let (mut old_t, mut t) = (0, 1);
    let mut count = 0;
    println!("\tEncontrando el xgcd({}): {} {} {}", count, old_r, old_t, old_s);
    while r != 0 {
        let quotient = old_r / r;
        let tmp = r;
        r = old_r - quotient * r;
        old_r = tmp;
        let tmp = s;
        s = old_s - quotient * s;
        old_s = tmp;
        let tmp = t;
        t = old_t - quotient * t;
        old_t = tmp;
        count += 1;
    }
    return (old_s, old_t, old_r); // (a, b, g)
}


fn main() {
    let prime = 13;

    println!("Operadores matemáticos en campo finito:");

    ////////////////////
    // FF Suma
    ////////////////////
    let lefthand = 4;
    let righthand = 12;
    println!(
        "\tSuma: ({} + {}) % {} = {}\n",
        lefthand,
        righthand,
        prime,
        (lefthand + righthand) % prime
    );

    ////////////////////
    // FF Resta
    ////////////////////
    let lefthand = 35;
    let righthand = 5;
    println!(
        "\tResta: ({} - {}) % {} = {}\n",
        lefthand,
        righthand,
        prime,
        (lefthand + righthand * -1 + prime) % prime
    );

    ////////////////////
    // FF Multiplicación
    ////////////////////
    let lefthand = 90;
    let righthand = 10;
    println!(
        "\tMultiplicación: ({} * {}) % {} = {}\n",
        lefthand,
        righthand,
        prime,
        (lefthand * righthand) % prime
    );

    ////////////////////
    // FF División
    ////////////////////
    let dividend = 2;
    let divisor = 3;
    println!("\tDivisión: ({} / {})", dividend, divisor);
    let (a, b, _) = xgcd(divisor, prime);
    println!(
        "\tse encontró: {}*{:2} + {}*{:2} = 1\n",
        a, divisor, b, prime
    );
    println!(
        "\t({} / {}) % {} = {}\n",
        dividend,
        divisor,
        prime,
        ((dividend * a % prime) + prime) % prime
    );
}
